src/edge_utils.c

```diff
@@ -2288,92 +2288,137 @@ static int routectl(int cmd, int flags, n2n_route_t *route, int if_idx) {
-       * See https://github.com/zerotier/ZeroTierOne/issues/178#issuecomment-204599227
-       */
-      n2n_sock_t sn;
-      n2n_route_t custom_route;
-      uint32_t *a;
+    return  edge_init_routes_linux(eee, routes, num_routes);
+#endif
 
-      if(eee->sn_route_to_clean) {
-	traceEvent(TRACE_ERROR, "Only one default gateway route allowed");
-	return(-1);
-      }
+#ifdef WIN32
+  return  edge_init_routes_win(eee, routes, num_routes);
+#endif
+}
 
-      if(eee->conf.sn_num != 1) {
-	traceEvent(TRACE_ERROR, "Only one supernode supported with routes");
-	return(-1);
-      }
+static int edge_init_routes_linux(n2n_edge_t *eee, n2n_route_t *routes, uint16_t num_routes) {
+#ifdef __linux__
+    int i;
+    for (i = 0; i<num_routes; i++) {
+        n2n_route_t *route = &routes[i];
+
+        if ((route->net_addr == 0) && (route->net_bitlen == 0)) {
+            /* This is a default gateway rule. We need to:
+            *
+            *  1. Add a route to the supernode via the host internet gateway
+            *  2. Add the new default gateway route
+            *
+            * Instead of modifying the system default gateway, we use the trick
+            * of adding a route to the networks 0.0.0.0/1 and 128.0.0.0/1, thus
+            * covering the whole IPv4 range. Such routes in linux take precedence
+            * over the default gateway (0.0.0.0/0) since are more specific.
+            * This leaves the default gateway unchanged so that after n2n is
+            * stopped the cleanup is easier.
+            * See https://github.com/zerotier/ZeroTierOne/issues/178#issuecomment-204599227
+            */
+            n2n_sock_t sn;
+            n2n_route_t custom_route;
+            uint32_t *a;
+
+            if (eee->sn_route_to_clean) {
+                traceEvent(TRACE_ERROR, "Only one default gateway route allowed");
+                return(-1);
+            }
 
-      if(supernode2addr(&sn, eee->conf.sn_ip_array[0]) < 0)
-	return(-1);
+            if (eee->conf.sn_num != 1) {
+                traceEvent(TRACE_ERROR, "Only one supernode supported with routes");
+                return(-1);
+            }
 
-      if(sn.family != AF_INET) {
-	traceEvent(TRACE_ERROR, "Only IPv4 routes supported");
-	return(-1);
-      }
+            if (supernode2addr(&sn, eee->conf.sn_ip_array[0]) < 0)
+                return(-1);
 
-      a = (u_int32_t*)sn.addr.v4;
-      custom_route.net_addr = *a;
-      custom_route.net_bitlen = 32;
-      custom_route.gateway = get_gateway_ip();
+            if (sn.family != AF_INET) {
+                traceEvent(TRACE_ERROR, "Only IPv4 routes supported");
+                return(-1);
+            }
 
-      if(!custom_route.gateway) {
-	traceEvent(TRACE_ERROR, "could not determine the gateway IP address");
-	return(-1);
-      }
+            a = (u_int32_t*)sn.addr.v4;
+            custom_route.net_addr = *a;
+            custom_route.net_bitlen = 32;
+            custom_route.gateway = get_gateway_ip();
```

Comment:
Windows can get default gateway from the routing table, the related function "GetIpForwardTable".